Mathlib/Topology/CWComplex/Classical/Finite.lean

```diff
@@ -0,0 +1,358 @@
+    (cell : (n : ℕ) → Type u) (map : (n : ℕ)  → (i : cell n) → PartialEquiv (Fin n → ℝ) X)
+    (eventually_isEmpty_cell : ∀ᶠ n in Filter.atTop, IsEmpty (cell n))
+    (finite_cell : ∀ (n : ℕ), _root_.Finite (cell n))
+    (source_eq : ∀ (n : ℕ) (i : cell n), (map n i).source = ball 0 1)
+    (continuousOn : ∀ (n : ℕ) (i : cell n), ContinuousOn (map n i) (closedBall 0 1))
+    (continuousOn_symm : ∀ (n : ℕ) (i : cell n), ContinuousOn (map n i).symm (map n i).target)
+    (pairwiseDisjoint' :
+      (univ : Set (Σ n, cell n)).PairwiseDisjoint (fun ni ↦ map ni.1 ni.2 '' ball 0 1))
+    (mapsTo : ∀ (n : ℕ) (i : cell n),
+      MapsTo (map n i) (sphere 0 1) (⋃ (m < n) (j : cell m), map m j '' closedBall 0 1))
+    (union' : ⋃ (n : ℕ) (j : cell n), map n j '' closedBall 0 1 = C) :
+    letI := mkFinite C cell map eventually_isEmpty_cell finite_cell source_eq continuousOn
+      continuousOn_symm pairwiseDisjoint' mapsTo union'
+    Finite C :=
+  letI := mkFinite C cell map eventually_isEmpty_cell finite_cell source_eq continuousOn
+      continuousOn_symm pairwiseDisjoint' mapsTo union'
+  { eventually_isEmpty_cell := eventually_isEmpty_cell
+    finite_cell := finite_cell}
+
+variable {X : Type*} [TopologicalSpace X] {C D : Set X} [RelCWComplex C D]
+
+lemma RelCWComplex.finite_of_finite_cells (finite : _root_.Finite (Σ n, cell C n)) : Finite C where
+  eventually_isEmpty_cell := by
+    simp only [Filter.eventually_atTop, ge_iff_le]
+    by_cases h : IsEmpty (Σ n, cell C n)
+    · use 0
+      intro n _
+      rw [isEmpty_sigma] at h
+      exact h n
+    -- We take the greatest `n` such that there is a `j : cell C n` and show that this fulfills
+    -- the necessary conditions.
+    rw [not_isEmpty_iff] at h
+    have _ := Fintype.ofFinite (Σ n, cell C n)
+    classical
+    let A := (Finset.univ : Finset (Σ n, cell C n)).image Sigma.fst
+    use A.max' (Finset.image_nonempty.2 Finset.univ_nonempty) + 1
+    intro m _
+    by_contra h'
+    have hmA : m ∈ A := by
+      simp only [Finset.mem_image, Finset.mem_univ, true_and, A]
+      simp only [not_isEmpty_iff, ← exists_true_iff_nonempty] at h'
+      obtain ⟨j, _⟩ := h'
+      use ⟨m, j⟩
+    linarith [A.le_max' m hmA]
+  finite_cell := fun _ ↦ Finite.of_injective (Sigma.mk _) sigma_mk_injective
+
+lemma RelCWComplex.finite_cells_of_finite [finite : Finite C] : _root_.Finite (Σ n, cell C n) := by
+  -- We show that there is a bijection between `Σ n, cell C n` and
+  -- `Σ (m : {m : ℕ // m < n}), cell C m`.
+  have h := finite.eventually_isEmpty_cell
+  have _ := finite.finite_cell
+  simp only [Filter.eventually_atTop, ge_iff_le] at h
+  rcases h with ⟨n, hn⟩
+  have : ∀ m (j : cell C m), m < n := by
+    intro m j
+    by_contra h
+    exact (hn m (not_lt.1 h)).false j
+  let f : (Σ (m : {m : ℕ // m < n}), cell C m) ≃ Σ m, cell C m := {
+    toFun := fun ⟨m, j⟩ ↦ ⟨m, j⟩
+    invFun := fun ⟨m, j⟩ ↦ ⟨⟨m, this m j⟩, j⟩
+    left_inv := by simp only [Function.LeftInverse, implies_true]
```

Comment:
This is a terminal simp: it's a bit more maintainable to not squeeze that, i.e. write `simp [Function.LeftInverse, implies_true]` instead. Same for the next line (only `Sigma.eta` can be removed).